### DISCUSSION TITLE
Proper way to test of nlmixr2 packages are in sync on CRAN

### DISCUSSION THREAD

**Discussion by john-harrold** [nlmixr2/nlmixr2#197] [Q&A] [answered]
I'm running into a problem where I have a package that suggests on some nlmixr2 packages. Specifically nlmixr2lib, nonmem2rx, and rxode2. Now I was testing to make sure they were installed, and if they are then different tests and examples will be run. 

I think the problem I'm running into is that when you pushed the new updates it put the set of packages in to a state where they are out of sync. So my tests say they are all installed so it runs, then I get failures because they are out of sync. Is there a way to "test" if they are installed and in sync?

**Comment by mattfidler** [comment]
I don't know of anything to test if they are installed and in sync (except if it fails it may be out of sync).

I would say that in 3-4 days things should stabilize.  The last binary packages were submitted to CRAN today so hopefully it will stabilize soon.

**Comment by john-harrold** [comment]
This isn't really just an inconvenience. When things are out of sync my examples and tests fail. So CRAN invites me to resubmit my package (or be kicked off :)). I stripped the nlmixr2 stuff out of ruminate and resubmitted so it wouldn't get kicked off of CRAN. It's not a big deal now because those pieces were in an experimental Shiny module. But I don't think this is feasible long term.

I should add that I think this will be a problem any package that suggests certain nlmxir2 packages will encounter in the future.

**Reply by mattfidler** [reply]
Hi John,

I know it is frustrating

If you use Suggests instead of imports CRAN in the past hasn't tried to kick packages off.

I cannot figure out a way to do a binary linkage without this painful step.  If you know a way, I am open to suggestions

**Comment by john-harrold** [comment]
Is there something simple I can run that will error out when things are out of sync? I can do that in a try catch environment, trap the error and use that as a way to determine if things are OK. I can do that once on package startup and then dump that status to a file in tempdir()?

**Reply by mattfidler** [reply, marked as answer]
Actually this is easier than I thought, just us the `Suggests` and then 

```r
Use `requireNamespace("nlmixr2", quietly = TRUE)` to test if package is installed
```


If it is out of sync, it will not be installed  and these tests would be skipped.

**Reply by john-harrold** [reply]
Currently I'm using system.file to determine if packages are installed. I'm getting values indicated they are installed but getting errors when I run tests/examples. Would this work differently?

**Reply by mattfidler** [reply]
Yes.  It loads it so it makes sure everything works.

**Reply by mattfidler** [reply]
This is not a nlmixr2 issue but a new CRAN check.  I had similar problems with a few of my packages. They do not install the Suggested packages or make them unavailable somehow and if your packages fail without them they will kick your package off of CRAN.

**Reply by john-harrold** [reply]
Yeah it passes that check. Because when they are missing the package can deal with this. The problem is when the packages are all there but out of sync. When this occurs my tests/examples are run and then fail because they are out of sync.

**Reply by mattfidler** [reply]
It should still fail to load the package if they are out of sync.

**Reply by john-harrold** [reply]
Does requireNamespace violate CRAN policies?

**Reply by john-harrold** [reply]
Hadley says it's OK:
https://r-pkgs.org/dependencies-mindset-background.html#sec-dependencies-namespace

So I think it's probably good :)

<img width="786" alt="image" src="https://github.com/nlmixr2/nlmixr2/assets/3170004/892f2446-c0ed-4274-989f-ebbeb3465df9">

**Reply by mattfidler** [reply]
Here is an artificial example when it is out of sync:

``` r
library(nlmixr2)
#> Loading required package: nlmixr2data
#> Error: package or namespace load failed for 'nlmixr2' in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]):
#>  namespace 'rxode2parse' 2.0.14 is being loaded, but >= 2.0.17 is required

requireNamespace("nlmixr2", quietly=TRUE)

print(requireNamespace("nlmixr2", quietly=TRUE))
#> [1] FALSE
```

<sup>Created on 2023-12-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by mattfidler** [reply]
This means that even if it errors on startup (when the package exists), it still can't require the namespace and returns `FALSE` so it  does work as intended.
